Execute simple JUnit tests and @QuarkusComponentTest first

gsmet opened the pull request:
I was hoping to constrain JUnit tests and @QuarkusComponentTests but @QuarkusComponentTest is in another artifact so this will have to do.

Fixes #46035.

Note that you can still break things if you specify another priority but that's how it is :).

gsmet committed:
Execute simple JUnit tests and @QuarkusComponentTest first

I was hoping to constrain JUnit tests and @QuarkusComponentTests but
@QuarkusComponentTest is in another artifact so this will have to do.

Related to #46035.

Note that you can still break things if you specify another priority but
that's how it is :).

geoand commented:
Rebased on main since the errors were weird and I want to see if they were actually caused by the PR

gsmet commented:
Yeah, I need to check what's going on. Will see if I can get it before we branch 3.19 but I'm not sure about that.

gsmet committed:
Properly isolate the ITs in a Failsafe execution

The ITs were run as part of the Surefire execution and apparently
something is leaking from them.
It wasn't an issue before because they were executed at the end of the
run but they were executed first with the new orderer given they are
not annotated with a Quarkus test annotation.

gsmet commented:
@geoand so there's something leaking somewhere but the test infra of the module was all borked. I isolated the ITs in a proper Failsafe run instead of grouping them with the Surefire tests and things are fine now.

The problem was caused by the leaking tests now being executed first as they are ITs and not having any sort of Quarkus test annotation.

At some point, we should probably try to figure out what the leak is but given Holly's ongoing work, I'm not sure it's worth it.

holly-cummins commented:
I've had a quick peek at this and #46035. I think: 

- In my new code, the 'last' app isn't closed, same as now. So I don't think #46035 would have worked with my branch before @gsmet's fix
- I think there's a chance that my new code would give better isolation to the ITs that Guillaume had to isolate, but I haven't actually checked 
- I have a suspicion this change is responsible for an ecosystem CI failure in the Pact extension, https://github.com/quarkiverse/quarkiverse/issues/94#issuecomment-2652807151; there was already some cross talk between tests and [super-fragile name-based ordering](https://github.com/quarkiverse/quarkus-pact/blob/main/quarkus-pact-provider/integration-tests/src/test/java/io/quarkiverse/pact/it/TheTestModeContractTestIT.java#L26-L31) in the Pact suite, and this probably perturbed that. Weirdly, the symptom is a classloading error, and I have no idea why test cross-talk can cause a classloading error, but I've been successfully ignoring the problem for some time. :)
